### PR TITLE
[air][tune] replace node:<ip> custom resource with NodeAffinitySchedulingPolicy

### LINKED ISSUE #31050: [train/tune] deprecate node:ip resources and use NodeAffinitySchedulingPolicy instead
### What happened + What you expected to happen

let's remove node:ip resources and use NodeAffinitySchedulingPolicy to schedule Actor associate to a node

master issue: https://github.com/ray-project/ray/issues/31038

### Versions / Dependencies

latest

### Reproduction script

n/a

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by matthewdeng**
Signed-off-by: Matthew Deng <matt@anyscale.com>

This PR changes usages of the `node:<ip>` custom resource as determined by querying [file:(air|tune|train).*\.py node:](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/ray-project/ray%24+file:%28air%7Ctune%7Ctrain%29.*%5C.py+node:). 

This is being used for:
- Collocating tasks (`_force_on_current_node`).
- Syncing files to specific IP addresses.
- Syncing files to _all_ other nodes.

## Why are these changes needed?

`node:<ip>` custom resources are being deprecated in Ray Core as described in https://github.com/ray-project/ray/issues/31038.

Changes:
1. Move `ray/tune/utils/node.py` to `ray/air/util/node.py`.
2. Create new `_get_node_id_from_node_ip` utility function.
3. Create new `_force_on_node` utility function which uses `NodeAffinitySchedulingStrategy`.
4. Update `_force_on_current_node` to use `_force_on_node`.
5. Update instances of `node:<ip>` custom resources to use the new utility functions.

## Related issue number

Closes #31050.

## Checks

Release tests run here: [`name:(tune_cloud.*|air_benchmark_tune_torch.*|air_benchmark_tensorflow.*)`](https://buildkite.com/ray-project/release-tests-pr/builds/26991#018605e4-7568-4913-8437-0b2bdb7d3af6).

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by rkooo567 — APPROVED** [review]
as long as it passes..

File: python/ray/air/util/node.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+def _get_node_id_from_node_ip(node_ip: str):
+    """Returns the node ID for the first alive node with the input IP."""
+    for node in ray.nodes():
+        if node["Alive"] and node["NodeManagerAddress"] == node_ip:
+            return node["NodeID"]
+
+    return None
+
+
+def _force_on_current_node(task_or_actor=None):
```
Comment (krfricke) [inline python/ray/air/util/node.py:13]:
```suggestion
def _force_on_current_node(task_or_actor: Optional[Union[ray.ObjectRef, ray.actor.ActorClass]] = None):
```

File: python/ray/tune/utils/file_transfer.py [hunk 2]
```diff
@@ -217,10 +233,15 @@ def delete_on_node(
...
+    node_id = _get_node_id_from_node_ip(node_ip)
+
     delete_task = _remote_delete_path.options(
         num_cpus=0,
-        resources={f"node:{node_ip}": 0.01},
-        scheduling_strategy=PlacementGroupSchedulingStrategy(placement_group=None),
+        scheduling_strategy=NodeAffinitySchedulingStrategy(
+            node_id=node_id,
+            soft=False,
+        ),
```
Comment (krfricke) [inline python/ray/tune/utils/file_transfer.py:244]: Should we also have a utility function for `force_on_node_ip/id`?

**Review by krfricke — APPROVED** [review]
Great, thanks!

**Review by Yard1 — APPROVED** [review]

File: python/ray/air/util/node.py (hunk 1 above)
Comment (matthewdeng) [reply python/ray/air/util/node.py]: Updated, though I believe the first should actually be a `ray.remote_function.RemoteFunction` - updated the arg name for clarity as well.

File: python/ray/tune/utils/file_transfer.py (hunk 2 above)
Comment (matthewdeng) [reply python/ray/tune/utils/file_transfer.py]: Makes sense! I added a `_force_on_node` that the caller can use in combination with `_get_node_id_from_node_ip`.

File: python/ray/air/util/node.py (hunk 1 above)
Comment (krfricke) [reply python/ray/air/util/node.py]: Ah sorry, you're right!

File: python/ray/air/util/node.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+    return None
+
+
+def _force_on_node(
+    node_id: str,
+    remote_func_or_actor_class: Optional[
+        Union[ray.remote_function.RemoteFunction, ray.actor.ActorClass]
+    ] = None,
+) -> Optional[Union[ray.remote_function.RemoteFunction, ray.actor.ActorClass]]:
```
Comment (justinvyu) [inline python/ray/air/util/node.py:19]: Maybe we can also add a type alias for the 2nd one?

```suggestion
) -> Union[Dict, Optional[Union[ray.remote_function.RemoteFunction, ray.actor.ActorClass]]]:
```

File: python/ray/air/util/node.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        return options
+
+    return remote_func_or_actor_class.options(**options)
+
+
+def _force_on_current_node(
+    remote_func_or_actor_class: Optional[
+        Union[ray.remote_function.RemoteFunction, ray.actor.ActorClass]
+    ] = None
+) -> Optional[Union[ray.remote_function.RemoteFunction, ray.actor.ActorClass]]:
```
Comment (justinvyu) [inline python/ray/air/util/node.py:51]: Same as above.

```suggestion
) -> Union[Dict, Optional[Union[ray.remote_function.RemoteFunction, ray.actor.ActorClass]]]:
```

File: python/ray/air/util/node.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+    options = {"scheduling_strategy": scheduling_strategy}
+
+    if remote_func_or_actor_class is None:
+        return options
+
+    return remote_func_or_actor_class.options(**options)
+
+
+def _force_on_current_node(
```
Comment (justinvyu) [inline python/ray/air/util/node.py:47]: Should we expose this as `@DeveloperAPI`?

**Review by justinvyu — APPROVED** [review]
Looks good. Can we add a test for `_force_on_node` with a Tune multinode test? Don't have that much context on how that test is set up, but would be nice to have.

File: python/ray/air/util/node.py (hunk 3 above)
Comment (matthewdeng) [reply python/ray/air/util/node.py]: Oof, good catch.

File: python/ray/air/util/node.py (hunk 5 above)
Comment (matthewdeng) [reply python/ray/air/util/node.py:47]: I think we should keep this as a private API, at least in its current state. I could imagine this being upstreamed to Core, in which we could mark it as a Developer API.

**Review by krfricke — APPROVED** [review]

**Pull request merged**